File: src/main/java/net/flyingfishflash/loremlist/core/validation/ValidUuidSet.java
package net.flyingfishflash.loremlist.core.validation;

import jakarta.validation.Constraint;
import jakarta.validation.Payload;
import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/** Annotation that validates all elements in a collection are valid UUIDs. */
@Target({ElementType.FIELD, ElementType.METHOD, ElementType.PARAMETER, ElementType.TYPE_USE})
@Retention(RetentionPolicy.RUNTIME)
@Documented
@Constraint(validatedBy = UuidSetValidator.class)
public @interface ValidUuidSet {
  String message() default "The set of UUID's is invalid";

  Class<?>[] groups() default {};

  Class<? extends Payload>[] payload() default {};

  boolean allowEmpty() default true;
}

File: src/main/kotlin/net/flyingfishflash/loremlist/core/validation/ValidUuidSet.kt
package net.flyingfishflash.loremlist.core.validation

import jakarta.validation.Constraint
import jakarta.validation.Payload
import kotlin.reflect.KClass

/**
 * Annotation that validates all elements in a collection are valid UUIDs.
 */
@Target(
  AnnotationTarget.FIELD,
  AnnotationTarget.PROPERTY_GETTER,
  AnnotationTarget.VALUE_PARAMETER,
  AnnotationTarget.PROPERTY,
  AnnotationTarget.TYPE,
)
@Retention(AnnotationRetention.RUNTIME)
@MustBeDocumented
@Constraint(validatedBy = [UuidSetValidator::class])
annotation class ValidUuidSet(
  val message: String = "The set of UUID's is invalid",
  val groups: Array<KClass<*>> = [],
  val payload: Array<KClass<out Payload>> = [],
  val allowEmpty: Boolean = true,
)
